internal/rundata/dut.go

```diff
@@ -0,0 +1,251 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rundata
+
+import (
+	"context"
+	"encoding/json"
+	"regexp"
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/openconfig/featureprofiles/internal/components"
+	"github.com/openconfig/ondatra"
+	"github.com/openconfig/ondatra/binding"
+	"github.com/openconfig/ondatra/gnmi/oc"
+	"github.com/openconfig/ondatra/gnmi/oc/ocpath"
+	"github.com/openconfig/ygnmi/ygnmi"
+)
+
+// dutInfo retrieves the vendor, model, and OS version from the device from various
+// OpenConfig paths.
+type dutInfo struct {
+	vendor string
+	model  string
+	osver  string
+}
+
+// setFromComponentChassis sets dutInfo from the first component of type CHASSIS.
+//
+//   - vendor from mfg-name (Arista, Cisco).
+//   - model from either description (Cisco, Juniper) or part-no (Arista).
+//   - osver from software-version (Juniper).
+func (di *dutInfo) setFromComponentChassis(ctx context.Context, y components.Y) {
+	if di.vendor != "" && di.model != "" && di.osver != "" {
+		return // No-op if nothing needs to be set.
+	}
+
+	const chassisType = oc.PlatformTypes_OPENCONFIG_HARDWARE_COMPONENT_CHASSIS
+
+	chassisIDs, err := y.FindByType(ctx, chassisType)
+	if err != nil {
+		glog.Errorf("Could not find chassis: %v", err)
+		return
+	}
+
+	chassisID := chassisIDs[0]
```

Comment:
How confident are we that the first component in the list is the correct one?

Reply:
There is typically just one chassis, but any chassis will work. Does not have to be the first one.